### PR TITLE
[Unit Tests] Mapper and Algorithm

### PR DESCRIPTION
**Pull request by zirui-song-18**
### Description
This PR contains developed unit tests up to July 14 2025. It mainly focuses on `org.opensearch.neuralsearch.sparse.algorithm` and `org.opensearch.neuralsearch.sparse.mapper`. Certain classes in the other package are also involved as they are high-dependency classes. In addition, a `org.opensearch.neuralsearch.sparse.testsPrepareUtils` is also created inside `tests` because it could provide some essential objects needed during tests.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/neural-search/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test/java/org/opensearch/neuralsearch/sparse/SparseTokensFieldTests.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        assertFalse("Should return false for null field", SparseTokensField.isSparseField(field));
+    }
+
+    public void testIsSparseFieldWhenFieldContainsSparseAttribute() throws Exception {
+        testsPrepareUtils prepareHelper = new testsPrepareUtils();
+        FieldInfo mockField = prepareHelper.prepareKeyFieldInfo();
+        Map<String, String> attributes = new HashMap<>();
+        attributes.put(SPARSE_FIELD, "true");
+
+        Field attributesField = FieldInfo.class.getDeclaredField("attributes");
```
Comment (chishui) [inline src/test/java/org/opensearch/neuralsearch/sparse/SparseTokensFieldTests.java:42]: can't you directly call `mockField.putAttribute()`?

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    private List<BytesRef> terms;
+    private InMemoryKey.IndexKey key;
+
+    @Before
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        MockitoAnnotations.openMocks(this);
+
+        terms = Arrays.asList(new BytesRef("term1"), new BytesRef("term2"));
```
Comment (chishui) [inline src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:31]: it seems `terms` is not modified anywhere, define it a constant?

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        // Setup
+        List<BytesRef> originalTerms = Arrays.asList(new BytesRef("term1"), new BytesRef("term2"));
+
+        // Execute - create task with null mergeState to test constructor
+        BatchClusteringTask task = new BatchClusteringTask(originalTerms, key, 0.5f, 0.3f, 10, null, null);
+
+        // Verify task is created
+        assertNotNull("Task should be created successfully", task);
+
+        Field termsField = BatchClusteringTask.class.getDeclaredField("terms");
```
Comment (chishui) [inline src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:45]: Since we own this class BatchClusteringTask, it's fine to add a getter for terms and get it from task and avoid having these tedious code

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+        // Modify original terms to verify deep copy
+        originalTerms.get(0).bytes[0] = (byte) 'X';
+
+        // Verify task's terms remain unchanged (proving deep copy worked)
+        assertEquals("Task's first term should still be 'term1'", "term1", taskTerms.get(0).utf8ToString());
+        assertNotEquals("Original term should now be different", "term1", originalTerms.get(0).utf8ToString());
+    }
+
+    public void testTaskCreationWithDifferentParameters() throws Exception {
```
Comment (chishui) [inline src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:61]: this test doesn't seem to bring much value as one value is not different comparing to another value when it's simply assignment. It makes more sense if you test null terms and empty terms.

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                    assertNotNull(
+                            "Task should be created with parameters: " + summaryPruneRatio + ", " + clusterRatio + ", " + nPostings,
+                            task
+                    );
+                }
+            }
+        }
+    }
+
+    public void testGetWithNullMergeState() throws Exception {
```
Comment (chishui) [inline src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:80]: (it's not about this test) I think we could annotate member variable `mergeState` `NonNull`

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+        // Create BatchClusteringTask
+        BatchClusteringTask task = new BatchClusteringTask(terms, key, 0.5f, 0.3f, 10, mergeState, keyFieldInfo);
+
+        // Execute and examine the result
+        List<Pair<BytesRef, PostingClusters>> result = task.get();
+
+        // Verify the returned clusters
+        assertNotNull("Result should not be null", result);
+        System.out.println("Result: " + result);
```
Comment (chishui) [inline src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:105]: avoid `System.out.println`

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        FieldInfo keyFieldInfo = prepareHelper.prepareKeyFieldInfo();
+
+        // Create BatchClusteringTask
+        BatchClusteringTask task = new BatchClusteringTask(terms, key, 0.5f, 0.3f, 10, mergeState, keyFieldInfo);
+
+        // Execute and examine the result
+        List<Pair<BytesRef, PostingClusters>> result = task.get();
+
+        // Verify the returned clusters
+        assertNotNull("Result should not be null", result);
```
Comment (chishui) [inline src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:104]: verify size?

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        // Should trigger early quit schema to return an empty list
+        assertEquals("Should return an empty list", 0, result.size());
+    }
+
+    public void testTermsDeepCopyInGet() throws Exception {
```
Comment (chishui) [inline src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:144]: are these two functions `readFromCacheOfOldSegment` `readFromCacheOfMergedSegment` tested?

**Comment by yuye-aws** [comment]
Hi @chishui , I have reviewed these PRs last week. Can you help review it?

File: src/test/java/org/opensearch/neuralsearch/sparse/mapper/SparseMethodContextTests.java [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+    public void testSparseMethodContextConstructorWithIOException() throws IOException {
+        StreamInput mockInput = mock(StreamInput.class);
+        when(mockInput.readString()).thenThrow(new IOException("Simulated IO error"));
+
+        expectThrows(IOException.class, () -> { new SparseMethodContext(mockInput); });
+    }
+
+    public void testSparseMethodConstructorWithStreamInput() throws IOException {
+        StreamInput mockStreamInput = mock(StreamInput.class);
```
Comment (chishui) [inline src/test/java/org/opensearch/neuralsearch/sparse/mapper/SparseMethodContextTests.java:80]: you already defined mockStreamInput with `@Mock`, no need to mock it again

File: src/test/java/org/opensearch/neuralsearch/sparse/testsPrepareUtils.java [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+    }
+
+    public static Settings prepareIndexSettings() {
+        return Settings.builder().put("index.number_of_shards", 1).put("index.number_of_replicas", 0).build();
+    }
+
+    public static ContentPath prepareContentPath() {
+        return new ContentPath();
+    }
+}
```
[the remark this answers is not in the document]
Comment (zirui-song-18) [reply src/test/java/org/opensearch/neuralsearch/sparse/testsPrepareUtils.java:369]: Ack. Will also check other files.

File: src/test/java/org/opensearch/neuralsearch/sparse/testsPrepareUtils.java [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+                null,                      // knnVectorsReaders
+                maxDocs,
+                InfoStream.getDefault(),
+                Executors.newSingleThreadExecutor(),
+                false                      // needsIndexSort
+        );
+        return mergeState;
+    }
+
+    public static IndexableFieldType prepareMockIndexableFieldType() {
```
[the remark this answers is not in the document]
Comment (zirui-song-18) [reply src/test/java/org/opensearch/neuralsearch/sparse/testsPrepareUtils.java:273]: Just as we talked, we choose to use this way as it is a final class.

File: src/test/java/org/opensearch/neuralsearch/sparse/SparseTokensFieldTests.java (hunk 1 above)
Comment (zirui-song-18) [reply src/test/java/org/opensearch/neuralsearch/sparse/SparseTokensFieldTests.java:42]: Ack

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java (hunk 3 above)
Comment (zirui-song-18) [reply src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:45]: Ack. Added a `getter` inside `BatchClusteringTask`

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        Field termsField = BatchClusteringTask.class.getDeclaredField("terms");
+        termsField.setAccessible(true);
+        @SuppressWarnings("unchecked")
+        List<BytesRef> taskTerms = (List<BytesRef>) termsField.get(task);
+
+        // Verify deep copy by checking actual content
+        assertEquals("First term should be 'term1'", "term1", taskTerms.get(0).utf8ToString());
+
+        // Modify original terms to verify deep copy
+        originalTerms.get(0).bytes[0] = (byte) 'X';
```
[the remark this answers is not in the document]
Comment (zirui-song-18) [reply src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:54]: That indeed is a workable way, but need to verify multiple times especially in some complex recursive structures. I would like to keep here unchanged.

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java (hunk 6 above)
Comment (zirui-song-18) [reply src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:105]: Ack.

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java (hunk 8 above)
Comment (zirui-song-18) [reply src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:144]: No. Those RandomClustering part seems not hit. Do you have any suggestions about how to test them?

File: src/test/java/org/opensearch/neuralsearch/sparse/mapper/SparseMethodContextTests.java (hunk 9 above)
Comment (zirui-song-18) [reply src/test/java/org/opensearch/neuralsearch/sparse/mapper/SparseMethodContextTests.java:80]: Ack

File: src/test/java/org/opensearch/neuralsearch/sparse/testsPrepareUtils.java [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+                0,                             // vectorDimension
+                VectorEncoding.FLOAT32,        // vectorEncoding
+                VectorSimilarityFunction.EUCLIDEAN, // vectorSimilarityFunction
+                false,                         // softDeletesField
+                false                          // isParentField
+        );
+        return keyFieldInfo;
+    }
+
+    public static SegmentInfo prepareSegmentInfo() {
```
[the remark this answers is not in the document]
Comment (zirui-song-18) [reply src/test/java/org/opensearch/neuralsearch/sparse/testsPrepareUtils.java]: I changed your previous create SegmentInfo with my new prepareSegmentInfo function.

File: src/test/java/org/opensearch/neuralsearch/sparse/testsPrepareUtils.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (chishui) [reply src/test/java/org/opensearch/neuralsearch/sparse/testsPrepareUtils.java]: oh, I mean the function name here, it's better not add `Mock` as it might be confusing with mocked object

File: src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java (hunk 5 above)
Comment (chishui) [reply src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:59]: if we expect it to be nonnull, we can make it fails early by constraining it to be nonnull, it's not about how test cases test it

Comment (zirui-song-18) [reply src/test/java/org/opensearch/neuralsearch/sparse/algorithm/BatchClusteringTaskTests.java:59]: OK. If later needs to modify BatchClusteringTask class, we can set it nonnull.
